File: cleanup.py
from contextlib import AbstractContextManager, closing
from typing import Iterable


# TODO Mention @contextmanager
class Resource(AbstractContextManager):
    def __init__(self, name: str):
        print(f"open {name}")
        self.closed = False
        self.name = name

    # def __enter__(self):
    #     print(f"enter {self.name}")
    #     return self

    # def __del__(self):
    #     print(f"del {self.name}")
    #     self.close()

    def __exit__(self, *exc):
        self.close()

    def close(self):
        if not self.closed:
            print(f"close {self.name}")
            self.closed = True


def prep_out(out_name: str, *, prep_names: Iterable[str]) -> Resource:
    writer = Resource(out_name)
    try:
        for name in prep_names:
            # with closing(Resource(name)) as reader:
            with Resource(name) as reader:
                # raise Exception("WELCOME TO YOUR DOOM!!")
                print(f"use {name}")
    except:
        writer.close()
        raise
    # else:
    #     print("Hi!")
    return writer


def main():
    with prep_out("out", prep_names=["a", "b"]) as writer:
        print("use out")


if __name__ == "__main__":
    main()

File: cleanup.zig
const print = @import("std").debug.print;

const Resource = struct {
    pub fn init(name: []const u8) !Resource {
        print("open {}\n", .{name});
        return Resource{ .name = name };
    }

    pub fn deinit(self: Resource) void {
        print("close {}\n", .{self.name});
    }

    name: []const u8,
};

const Error = error{WelcomeToYourDoom};

fn prepOut(out_name: []const u8, prep_names: []const []const u8) !Resource {
    var writer = try Resource.init(out_name);
    errdefer writer.deinit();
    for (prep_names) |name| {
        var reader = try Resource.init(name);
        defer reader.deinit();
        // if (true) return Error.WelcomeToYourDoom;
        print("use {}\n", .{name});
    }
    return writer;
}

pub fn main() !void {
    var writer = try prepOut("out", &[_][]const u8{ "a", "b" });
    defer writer.deinit();
    print("use out\n", .{});
}
